zfit/core/minimizer.py

```diff
@@ -5,66 +5,214 @@
 
 import abc
 import collections
+from collections import OrderedDict
+import contextlib
 
 import numpy as np
 import tensorflow as tf
 import tensorflow_probability as tfp
 
 import zfit.core.math as zmath
 from zfit import ztf
-import zfit.ztf
+from zfit.minimizers.state import MinimizerState
 
 
-class AbstractMinimizer(object):
+class MinimizerInterface(object):
     """Define the minimizer interface."""
 
     @abc.abstractmethod
     def minimize(self):
         raise NotImplementedError
 
-    def _minimize(self, *args, **kwargs):
+    def _minimize(self):
         raise NotImplementedError
 
-    @abc.abstractmethod
-    def edm(self):
+    def _minimize_with_step(self):
         raise NotImplementedError
 
-    def _edm(self, *args, **kwargs):
+    @abc.abstractmethod
+    def edm(self):
         raise NotImplementedError
 
     @abc.abstractmethod
     def fmin(self):
         raise NotImplementedError
 
+    def step(self):
+        raise NotImplementedError
+
+    def _step_tf(self):
+        raise NotImplementedError
+
+    def _step(self):
+        raise NotImplementedError
+
     @property
     @abc.abstractmethod
     def tolerance(self):
         raise NotImplementedError
 
+    def _tolerance(self):
+        raise NotImplementedError
+
     @abc.abstractmethod
-    def status(self):
+    def get_state(self):
         raise NotImplementedError
 
+    @abc.abstractmethod
+    def hesse(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def error(self):
+        raise NotImplementedError
 
-class BaseMinimizer(object):
+    @abc.abstractmethod
+    def set_error_method(self, method):
+        raise NotImplementedError
 
-    def __init__(self, name="BaseMinimizer", tolerance=1e-8, sess=None, *args, **kwargs):
-        super(BaseMinimizer, self).__init__(*args, **kwargs)
-        self.name = name
+    @abc.abstractmethod
+    def set_error_options(self):
+        raise NotImplementedError
+
+
+class BaseMinimizer(MinimizerInterface):
 
-        self.sess = sess
+
+
+    def __init__(self, loss, parameters=None, tolerance=1e-8, name="BaseMinimizer", *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._current_error_method = self._error_methods.get('default')
+        self._current_error_options = {}
+        self._minimizer_state = MinimizerState()
+        self.name = name
         self.tolerance = tolerance
+        self._sess = None
+        self._parameters = OrderedDict()
+        self.loss = loss
+        self.set_parameters(parameters)
+
+    def __init_subclass__(cls, **kwargs):
+        cls._error_methods = {'default': None}  # TODO: set default error method
+
+    @property
+    def edm(self):
+        return self.get_state(copy=False).edm
+
+    @property
+    def fmin(self):
+        return self.get_state(copy=False).fmin
+
+    def hesse(self):
```

Comment:
I guess we can't just remove it? It would break with `abc`?

Reply:
"yes", as long as it is not overriden by the subclass. But why not provide a default implementation?
It's on the TODO ;)

Reply:
It's a philosophy thing: what does a "default" implementation mean? Each minimizer is going to be different, isn't it?

Reply:
Mmm, to be honest, I'm not sure. Maybe let's leave it like it is for now, then we can revisit if the need appears.

Reply:
> like now
-> with a default implementation?

Reply:
Yeah, or if nothing sensible can be made, raise a `NotImplementedError`.